Use ThreadSafeWeakPtr in ThreadGroup

Constellation opened the pull request:
#### 9b28b7aca993535511058d22d563e75dc50b3911
<pre>
Use ThreadSafeWeakPtr in ThreadGroup
<a href="https://bugs.webkit.org/show_bug.cgi?id=248103">https://bugs.webkit.org/show_bug.cgi?id=248103</a>
rdar://102528206

Reviewed by Darin Adler and Chris Dumez.

We were using std::shared_ptr and std::weak_ptr in ThreadGroup, and the reason of this is because of lack of thread-safety
with WeakPtr / Ref. But now, we have ThreadSafeWeakPtr, so let&apos;s use it instead.

* Source/JavaScriptCore/heap/MachineStackMarker.h:
* Source/WTF/wtf/ThreadGroup.h:
* Source/WTF/wtf/Threading.cpp:
(WTF::Thread::didExit):
(WTF::Thread::addToThreadGroup):
* Source/WTF/wtf/Threading.h:
* Source/WTF/wtf/linux/RealTimeThreads.h:
* Source/WTF/wtf/threads/Signals.cpp:
(WTF::activeThreads):
* Tools/TestWebKitAPI/Tests/WTF/ThreadGroup.cpp:
(TestWebKitAPI::testThreadGroup):
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/257595@main">https://commits.webkit.org/257595@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e2749687c36be0b369a0b6c2ec951d4f70cdf6dd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99353 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8554 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32476 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108738 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168989 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/103350 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/9115 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85873 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91844 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106666 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105111 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/6894 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90436 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33875 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88713 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21783 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/76751 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/90048 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2424 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23300 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/85889 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2327 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/45702 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28881 "Passed tests") | 
| | [💥 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/8488 "An unexpected error occured. Recent messages:Failed to print configuration") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42775 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/88756 "Built successfully") | 
| [❌ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2670 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4195 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19849 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/WTF/wtf/Threading.h:
```diff
@@ -367,7 +368,7 @@ class WTF_CAPABILITY("is current") Thread : public ThreadSafeRefCounted<Thread>,
     WordLock m_mutex;
     StackBounds m_stack { StackBounds::emptyBounds() };
-    HashMap<ThreadGroup*, std::weak_ptr<ThreadGroup>> m_threadGroupMap;
+    HashMap<ThreadGroup*, ThreadSafeWeakPtr<ThreadGroup>> m_threadGroupMap;
```
This should be WeakHashSet once `WeakHashSet` supports ThreadSafeWeakPtr in the future.

cdumez reviewed (approved):
LGTM

darinadler commented on Tools/TestWebKitAPI/Tests/WTF/ThreadGroup.cpp:
```diff
@@ -101,7 +101,7 @@ TEST(WTF, ThreadGroupAddCurrentThread)
 TEST(WTF, ThreadGroupDoNotAddDeadThread)
 {
-    std::shared_ptr<ThreadGroup> threadGroup = ThreadGroup::create();
+    Ref<ThreadGroup> threadGroup = ThreadGroup::create();
```
`auto`

Constellation replied:
Changed.

darinadler commented on Tools/TestWebKitAPI/Tests/WTF/ThreadGroup.cpp:
```diff
@@ -115,7 +115,7 @@ TEST(WTF, ThreadGroupAddDuplicateThreads)
     Lock lock;
     Condition restartCondition;
-    std::shared_ptr<ThreadGroup> threadGroup = ThreadGroup::create();
+    Ref<ThreadGroup> threadGroup = ThreadGroup::create();
```
`auto`

Constellation replied:
Changed.

darinadler commented on Tools/TestWebKitAPI/Tests/WTF/ThreadGroup.cpp:
```diff
@@ -163,7 +163,7 @@ TEST(WTF, ThreadGroupRemove)
     Vector<Ref<Thread>> threads;
 
-    auto threadGroup = ThreadGroup::create();
+    RefPtr<ThreadGroup> threadGroup = ThreadGroup::create();
```
Why change from `auto`?

Constellation replied:
This needs to be RefPtr to allow clearing this with `= nullptr`.

darinadler replied:
For the long term we should fix things so we can just write `RefPtr` here instead of `RefPtr<ThreadGroup>`. Might need a new deduction guide that is not already present in `RefPtr.h` or `Ref.h`.

darinadler commented on Source/WTF/wtf/threads/Signals.cpp:
```diff
@@ -297,13 +297,13 @@ inline void setExceptionPorts(const AbstractLocker& threadGroupLocker, Thread& t
         activeThreads.construct(ThreadGroup::create());
     });
-    return (*activeThreads.get());
+    return activeThreads.get().get();
```
Don’t need the second `get()`.

Constellation replied:
Changed.

darinadler commented on Source/WTF/wtf/ThreadGroup.h:
```diff
@@ -58,10 +58,6 @@ class ThreadGroup final : public std::enable_shared_from_this<ThreadGroup> {
-    {
-        return shared_from_this();
-    }
 
```
Stray blank line. Remove it?

Constellation replied:
Removed.

darinadler commented on Source/WTF/wtf/ThreadGroup.h:
```diff
@@ -34,15 +34,15 @@ namespace WTF {
+    static Ref<ThreadGroup> create()
     {
-        return std::allocate_shared<ThreadGroup>(FastAllocator<ThreadGroup>());
+        return adoptRef(*new ThreadGroup());
```
I would take out the parentheses after ThreadGroup. Also, should move this to the .cpp file since this is probably not good to inline. We’d want to inline the constructor instead.

Constellation replied:
OK, changed.

darinadler commented on Tools/TestWebKitAPI/Tests/WTF/ThreadGroup.cpp:
```diff
@@ -34,7 +34,7 @@ namespace TestWebKitAPI {
 static void testThreadGroup(Mode mode)
 {
-    std::shared_ptr<ThreadGroup> threadGroup = ThreadGroup::create();
+    Ref<ThreadGroup> threadGroup = ThreadGroup::create();
```
Should be able to use `auto` here.

Constellation replied:
Changed.

Constellation committed:
Use ThreadSafeWeakPtr in ThreadGroup
https://bugs.webkit.org/show_bug.cgi?id=248103
rdar://102528206

Reviewed by Darin Adler and Chris Dumez.

We were using std::shared_ptr and std::weak_ptr in ThreadGroup, and the reason of this is because of lack of thread-safety
with WeakPtr / Ref. But now, we have ThreadSafeWeakPtr, so let's use it instead.

* Source/JavaScriptCore/heap/MachineStackMarker.h:
* Source/WTF/wtf/ThreadGroup.h:
* Source/WTF/wtf/Threading.cpp:
(WTF::Thread::didExit):
(WTF::Thread::addToThreadGroup):
* Source/WTF/wtf/Threading.h:
* Source/WTF/wtf/linux/RealTimeThreads.h:
* Source/WTF/wtf/threads/Signals.cpp:
(WTF::activeThreads):
* Tools/TestWebKitAPI/Tests/WTF/ThreadGroup.cpp:
(TestWebKitAPI::testThreadGroup):
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/257595@main
